frontend/cs/scs/builder.go

```diff
@@ -686,3 +697,39 @@ func (builder *builder) ToCanonicalVariable(v frontend.Variable) frontend.Canoni
 		return term
 	}
 }
+
+func (builder *builder) storeWireLocation(a, b int) {
+	for i, v := range []int{a, b} {
+		if _, ok := builder.mRangecheckInterleave[i][v]; !ok {
+			builder.mRangecheckInterleave[i][v] = builder.cs.GetNbConstraints()
+		}
+	}
+}
+
+func (builder *builder) GetWireGates(wires []frontend.Variable) [2][]int {
```

Comment:
I believe one alternative way to implement this would be to not store additional tracking data for the wires, but instead do **ALL** the extra computation within that function.

This would be done by scraping the wire IDs of `wires` within a set : `map[int]struct{}`. Then, we could try to iterate on all the `xA` and then all the `xB` of the compiled circuit to try and find all the wires in the set. The advantage of that solution is that we only "pay the cost" of `GetWireGates` when we use it; meaning that the compilation performances for the other circuits are independent.

Another point, is that if we try to find all the wires within xA (and luckily succeed) then it simplifies the range-checking mechanism later on the wizard because, then, we only have to look at a one column instead of 2. (I genuinely don't know if that would work though)

Reply:
I checked once if we could only get away with being able to check only a single column, but this doesn't always work - it depends on the order of arguments when calling `api.Mul`. Considering that the complexity may be quite high (ECDSA -> scalarmul -> field emulation -> defer -> api.Mul), then it is difficult to reliably ensure we only assign range checkable wire to the first or second position.

Reply:
But I guess one more thing what we can do is to implement `frontend.Checker` on gnark side -- maybe then we have more information? Currently we implement it on zkevm side (by appending all the variables we want to range check) and then use `GetWireGates` to get the corresponding gate locations. But if we implement it on gnark side and finding the first gate location then it is easier (as the number of gates may be smaller). But this makes the compilation slower (but only when we have the builder which implements `frontend.Checker`).

Reply:
You mean, moving the `externalRangeChecker` into `gnark`?

Reply:
> I checked once if we could only get away with being able to check only a single column, but this doesn't always work - it depends on the order of arguments when calling api.Mul. Considering that the complexity may be quite high (ECDSA -> scalarmul -> field emulation -> defer -> api.Mul), then it is difficult to reliably ensure we only assign range checkable wire to the first or second position.

I think it is fine if we go with an optimistic approach. e.g. first try to find the wire in xA. If not, then too bad : try to find it in xB.

Reply:
Yes, and then allowing to output something what the user needs. For example, which gates should be range checking gates etc.

Reply:
Yep. But after thinking a few minutes about it, I'm not sure it can be nicely done. I guess it would be better to try the memoryless approach where we perform searches only at call to `GetWireGate` and do not store the first-gate-per-wire map.

Reply:
Now performs all computation when calling `GetWireGates`